File: BackEnd-Siukat/routes/verifikasi.go
package routes

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/middlewares"
	"BackEnd-Siukat/models"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
)

func VerifikasiRoutes(r *gin.RouterGroup) {
	group := r.Group("/verifikasi")
	group.Use(middlewares.JwtAuth())

	group.GET("", func(c *gin.Context) {
		noPeserta := c.GetString("no_peserta")
		if noPeserta == "" {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "No peserta not found in token"})
			return
		}

		tx := config.DB.Begin()
		defer func() {
			if r := recover(); r != nil {
				tx.Rollback()
			}
		}()

		var info models.Info
		if err := tx.First(&info).Error; err != nil {
			tx.Rollback()
			c.JSON(http.StatusInternalServerError, err)
			return
		}

		var verifikasi models.Verifikasi
		// Legacy: req.user.no_peserta is used. In Verifikasi model, no_peserta is int.
		noPesertaInt, _ := strconv.Atoi(noPeserta)
		if err := tx.Where("no_peserta = ?", noPesertaInt).First(&verifikasi).Error; err != nil {
			tx.Rollback()
			c.JSON(http.StatusInternalServerError, err)
			return
		}

		var cmahasiswa models.CMahasiswa
		if err := tx.Where("no_peserta = ?", noPeserta).First(&cmahasiswa).Error; err != nil {
			tx.Rollback()
			c.JSON(http.StatusInternalServerError, err)
			return
		}

		tx.Commit()

		data := gin.H{
			"info":             info,
			"verifikasi":       verifikasi,
			"result_bidikmisi": verifikasi.VerBidikMisi,
			"result_kipk":     verifikasi.VerKipk,
			"result_kjmu":     verifikasi.VerKjmu,
		}

		switch info.Stage {
		case "snmptn":
			fakultasStr := strconv.Itoa(cmahasiswa.FakultasCmahasiswa)
			if fakultasStr == "16" || cmahasiswa.ProdiCmahasiswa == "12076" || cmahasiswa.ProdiCmahasiswa == "12086" || cmahasiswa.ProdiCmahasiswa == "12066" {
				if verifikasi.VerAkademik == "lolos" && verifikasi.VerKeterampilan == "lolos" {
					data["result_akademik"] = "lolos"
				} else if verifikasi.VerAkademik == "tidak_lolos" || verifikasi.VerKeterampilan == "tidak_lolos" {
					data["result_akademik"] = "tidak_lolos"
				} else if verifikasi.VerAkademik == "belum_verifikasi" || verifikasi.VerKeterampilan == "belum_verifikasi" {
					data["result_akademik"] = "belum_verifikasi"
				}
			} else {
				data["result_akademik"] = verifikasi.VerAkademik
				data["result_bidikmisi"] = verifikasi.VerBidikMisi
				data["result_kipk"] = verifikasi.VerKipk
				data["result_kjmu"] = verifikasi.VerKjmu
			}
		case "sbmptn":
			data["result_akademik"] = "lolos"
		default:
			// Mandiri
			data["result_akademik"] = "lolos"
		}
		c.JSON(http.StatusOK, data)
	})
}

File: ZhammDataBeLama/siukat-service-dev_snmptn_2022/routes/verifikasi.js
var express = require("express");
var router = express.Router();
var db = require("../config/database");
var Verifikasi = require("../models/verifikasi");
var CMahasiswa = require("../models/cmahasiswa");
var Info = require("../models/info");
var passport = require("passport");

router.get("/", passport.authenticate("jwt", { session: false }), function (
    req,
    res
) {
    return db
        .transaction(function (t) {
            var data = {};
            Info.findOne()
                .then((info) => {
                    data.info = info;
                    return Verifikasi.findOne({
                        where: { no_peserta: req.user.no_peserta },
                    });
                })
                .then((verifikasi) => {
                    data.verifikasi = verifikasi;

                    return CMahasiswa.findOne({
                        where: { no_peserta: req.user.no_peserta },
                    });
                })
                .then((cmahasiswa) => {
                    data.result_bidikmisi = data.verifikasi.ver_bidik_misi;
                    data.result_kipk = data.verifikasi.ver_kipk;
                    data.result_kjmu = data.verifikasi.ver_kjmu;
                    if (data.info.stage === "snmptn") {
                        if (
                            cmahasiswa.fakultas_cmahasiswa.toString() === "16" ||
                            cmahasiswa.prodi_cmahasiswa.toString() === "12076" ||
                            cmahasiswa.prodi_cmahasiswa.toString() === "12086" ||
                            cmahasiswa.prodi_cmahasiswa.toString() === "12066"
                        ) {
                            if (
                                data.verifikasi.ver_akademik === "lolos" &&
                                data.verifikasi.ver_keterampilan === "lolos"
                            ) {
                                // Jika akademik dan keterampilan lolos, bidik misi mengikuti
                                data.result_akademik = "lolos";
                            } else if (
                                data.verifikasi.ver_akademik === "tidak_lolos" ||
                                data.verifikasi.ver_keterampilan === "tidak_lolos"
                            ) {
                                // Jika akademik atau keterampilan tidak lolos, maka bidik misi juga tidak lolos
                                data.result_akademik = "tidak_lolos";
                            } else if (
                                data.verifikasi.ver_akademik === "belum_verifikasi" ||
                                data.verifikasi.ver_keterampilan === "belum_verifikasi"
                            ) {
                                // Jika akademik atau keterampilan belum verifikasi, maka bidik misi mengikuti
                                data.result_akademik = "belum_verifikasi";
                            }
                        } else {
                            // Akademik tidak mempertimbangkan keterampilan (EQUAL)
                            data.result_akademik = data.verifikasi.ver_akademik;
                            data.result_bidikmisi = data.verifikasi.ver_bidik_misi;
                            data.result_kipk = data.verifikasi.ver_kipk;
                            data.result_kjmu = data.verifikasi.ver_kjmu;
                        }
                    } else if (data.info.stage === "sbmptn") {
                        // Semua lolos akademik, bidik misi mengikuti
                        data.result_akademik = "lolos";
                    } else {
                        // Mandiri
                        // Semua lolos akademik, bidik misi tidak ada
                        data.result_akademik = "lolos";
                    }
                    res.json(data);
                });
        })
        .catch(function (err) {
            res.status(500).json(err);
        });
});

module.exports = router;
